merino-suggest/src/providers/client_variant_filter.rs

```diff
@@ -0,0 +1,134 @@
+//! A suggestion provider switches between a matching and default provider based on the client variant string.
+use crate::{CacheInputs, SuggestError, SuggestionProvider, SuggestionRequest, SuggestionResponse};
+use async_trait::async_trait;
+
+/// A provider that gives suggestions base
+pub struct ClientVariantFilterProvider {
+    /// Provider to use for suggestions if there is a client variant match
+    matching_provider: Box<dyn SuggestionProvider>,
+    /// Provider to use for suggestions if there isn't a client variant match
+    default_provider: Box<dyn SuggestionProvider>,
+    /// String use to match with client variants from suggest requests
+    client_variant: String,
+}
+
+impl ClientVariantFilterProvider {
+    /// Create a boxed Client Variant Provider
+    pub fn new_boxed(
+        matching_provider: Box<dyn SuggestionProvider>,
+        default_provider: Box<dyn SuggestionProvider>,
+        client_variant: String,
+    ) -> Box<Self> {
+        Box::new(Self {
+            matching_provider,
+            default_provider,
+            client_variant,
+        })
+    }
+}
+
+#[async_trait]
+impl SuggestionProvider for ClientVariantFilterProvider {
+    fn name(&self) -> String {
+        format!(
+            "ClientVariant(matching:{}, default:{}, client_variant match: {})",
+            self.matching_provider.name(),
+            self.default_provider.name(),
+            self.client_variant,
+        )
+    }
+
+    async fn suggest(
+        &self,
+        request: SuggestionRequest,
+    ) -> Result<SuggestionResponse, SuggestError> {
+        let req = request.clone();
+        let provider = if req
+            .client_variants
+            .unwrap_or_default()
+            .contains(&self.client_variant)
```

Comment:
I realize that you can do the same `as_ref` trick here that you used in `cache_inputs`.